subsys/bluetooth/host/buf.c

```diff
@@ -100,9 +91,11 @@ struct net_buf *bt_buf_get_evt(uint8_t evt, bool discardable,
 	switch (evt) {
 #if defined(CONFIG_BT_CONN) || defined(CONFIG_BT_ISO)
 	case BT_HCI_EVT_NUM_COMPLETED_PACKETS:
+#endif /* CONFIG_BT_CONN || CONFIG_BT_ISO */
+	case BT_HCI_EVT_CMD_STATUS:
+	case BT_HCI_EVT_CMD_COMPLETE:
```

Comment:
BT Public APIs are callable from application's pre-emptive threads (main thread), does this change mandate that the HCI Driver's send interface (directly called from APIs) delegate the processing of HCI commands to a co-operative thread (previously the tx thread) to ensure the Controller's non-reentrant functions are not directly called from applications?

If the Controller's function is made re-entrant, then can there be a case where the Kernel switches contexts (time slice) between two threads that are calling APIs that lead to the HCI Driver's send interface (command buffer depth > 1) being invoked? In which case will a pool size of 1 be sufficient?

Example, a thread that is transmitting data, another thread that could issue HCI channel classification command or HCI vendor specific Tx Power command.

Can we instead ensure/reserve a buf for the command complete/status event before HCI Driver's send interface is invoked?

Reply:
> HCI Driver's send interface (directly called from APIs) delegate the processing of HCI commands to a co-operative thread

Already the case. The commands are all guaranteed to be sent through `bt_send()` on the system workqueue.

Reply:
here: https://github.com/jori-nordic/zephyr/blob/c6fd58e73d7b7375b143e131e070087bb7ee8b55/subsys/bluetooth/host/hci_core.c#L4681-L4697

Reply:
> > HCI Driver's send interface (directly called from APIs) delegate the processing of HCI commands to a co-operative thread
> 
> Already the case. The commands are all guaranteed to be sent through `bt_send()` on the system workqueue.

Great, at one point I was concern that LL interface where called from pre-emptive thread.

How is the case handled when scan disable requires that an adv report be processed before command complete is generated? (cross over of command and event). API call from system workqueue would deadlock waiting for command complete?

@rugeGerritsen isn't it a similar scenario with the deadlock in receiving disconnection event when data is being sent from system workqueue execution context?

Will the solution be to disallow API calls from system workqueue execution context?

Reply:
> How is the case handled when scan disable requires that an adv report be processed before command complete is generated? (cross over of command and event). API call from system workqueue would deadlock waiting for command complete?

Adv reports blocking the event stream is described in https://github.com/zephyrproject-rtos/zephyr/issues/50786. I'll be submitting a separate fix for this.

Reply:
@cvinayak regarding the re-entrancy of the LL, you could use something like the SDC, a simple 1-count semaphore: https://github.com/jori-nordic/sdk-nrf/blob/32b91a8fe62d6c2c86d076344c03cbb79f5e91b2/subsys/bluetooth/controller/hci_driver.c#L547-L552

Reply:
> > How is the case handled when scan disable requires that an adv report be processed before command complete is generated? (cross over of command and event). API call from system workqueue would deadlock waiting for command complete?
> 
> Adv reports blocking the event stream is described in #50786. I'll be submitting a separate fix for this.

Thanks for reminding it, forgot that one.



> @cvinayak regarding the re-entrancy of the LL, you could use something like the SDC, a simple 1-count semaphore: https://github.com/jori-nordic/sdk-nrf/blob/32b91a8fe62d6c2c86d076344c03cbb79f5e91b2/subsys/bluetooth/controller/hci_driver.c#L547-L552

ok... will study the latest data flow path and analyse based on the fact now that system work queue is the highest priority (value -1) co-operative thread from which LL interfaces are directly called compared to earlier Tx thread (-7), and wherein Rx Prio being at (-6) in Zephyr Controller's hci driver that received command complete/status/num completes/disconnection complete.

Now that system work queue is handling command complete/status/num complete in sync pool, has disconnection complete been analysed for sync pool use?

No further action for this PR.

Reply:
> Now that system work queue is handling command complete/status/num complete in sync pool, has disconnection complete been analysed for sync pool use?

Disconnect complete events cannot be put in the sync pool because their event buffers are retained beyond "prio" context. This has to be fixed first.

Reply:
> ok... will study the latest data flow path and analyse based on the fact now that system work queue is the highest priority (value -1) co-operative thread from which LL interfaces are directly called compared to earlier Tx thread (-7), and wherein Rx Prio being at (-6) in Zephyr Controller's hci driver that received command complete/status/num completes/disconnection complete.

Thanks :) I would prefer solutions that do no rely on thread priorities. If you see any issues, I think we should look into fixing it with explicit synchronization.